marconi-mamba/test/Spec/Marconi/Mamba/Api/Query/Indexers/Utxo.hs

```diff
@@ -136,3 +158,82 @@ generateUtxoRowJson = do
                 (Just $ C.hashScript script)
         utxoRow = UtxoRow utxo (C.SlotNo 1) blockHeaderHash
     pure $ Aeson.encodePretty utxoRow
+
+-- | Start a servant process and execute the client action
+withUserApp
+  :: IndexerEnv
+  -> (Warp.Port -> IO (JsonRpcResponse String UtxoQueryResult)) -- ^ hoisted JSON-RPC client such that it can run in IO
+  -> IO (JsonRpcResponse String UtxoQueryResult)
+withUserApp env  = Warp.testWithApplication(pure $ marconiApp env)
+
+-- | Create http JSON-RPC client, and execute the JsonRpc functions
+clientEnv
+  :: String                                       -- ^ bech32 ShelleyEra address to fetch Utxo's for
+  -> Warp.Port                                    -- ^ JSON-RPC port the server will respond to
+  -> IO (JsonRpcResponse String UtxoQueryResult)  -- ^ response receive from the JSON-RPC server
+clientEnv address port = do
+  manager' <- newManager defaultManagerSettings
+  let env = mkClientEnv manager' (baseUrl port)
+  let (rpcEcho :<|> rpcTargets :<|> rpcUtxos) = getClients env
+  -- test will fail if these calls/routes fail.
+  _ <- rpcEcho "client calling "  -- we just want to know if the route is alive
+  _ <- rpcTargets ""              -- we just want to know if the route is alive
+  rpcUtxos address
+
+-- | Test round trip Utxos thruough JSON-RPC http server
+-- We compare a represenation of the generated UtxoEvents
+-- with those fetched from the JSON-RPC  server. The purpose of this is:
+--   + RPC server routes the request to the correct handler
+--   + Events are serialized/deserialized thru the RPC layer and it various wrappers correctly
+
+jsonRpcTest :: Property
+jsonRpcTest =  property $ do
+  event <- forAll genEvent
+  env <- liftIO . UIQ.initializeEnv $ Nothing
+  let
+    callback :: Utxo.UtxoIndexer -> IO ()
+    callback = atomically . UIQ.writeTMVar' (env ^. uiIndexer) -- update the indexer
+  liftIO . mocUtxoWorker callback [event] $ Utxo.Depth 10
+  let (qAddresses :: [String]) =
+        Set.toList . Set.fromList . fmap (unpack . C.serialiseAddress)
+        . mapMaybe (addressAnyToShelley . Utxo._address)
+        . Set.toList . Utxo.ueUtxos
+        $ event
+  rpcResponses <- liftIO $ traverse (\a -> withUserApp env (clientEnv a) ) qAddresses
```

Comment:
Doesn't seem very efficient. This is starting a server on every queried address, no? I think this can be refactored to only call once the server initialisation.